tests/mdf/model_with_control.py

```diff
@@ -65,7 +65,7 @@
             },
             {
                 pnl.PROJECTIONS: (pnl.THRESHOLD, Decision),
-                pnl.ALLOCATION_SAMPLES: np.arange(0.1, 1.01, 0.3),
+                pnl.ALLOCATION_SAMPLES: (0.1, 0.4, 0.7000000000000001, 1.0000000000000002),
```

Comment:
If other things are held up by this PR, I might lean in favor of xfailing the failing tests in test_mdf.py instead. This change does make them pass, but the script generated would also crash if run outside of the test environment. This is because of the new `np.float64(`s in #3207 - the numpy import isn't added, but when run by the test, numpy is already imported, which is unintended. I'll need to spend some time reworking these tests, because there might be other preexisting issues unrelated to numpy 2.

Reply:
The fix in this PR is for the presence of nested parentheses. The issue exists irrespective of Numpy 2 and the modified version of this test fails also when using numpy-1.26.
Ideally, this test would use all possible expressions that can be passed to ALLCOATION_SAMPLES; list, arrange, np.array, tuple, ..., to maximize coverage that the test provides.

There might be other issues caused by Numpy 2 that are not caused by existing tests (either by passing accidentally, or not being tested at all), but I'd prefer not to postpone Numpy 2 enablement for those cases.

Reply:
>The issue exists irrespective of Numpy 2 and the modified version of this test fails also when using numpy-1.26.

Yeah, this is basically what I meant as justification for just xfailing it here instead. But I don't have major gripes with proceeding with this as-is

Reply:
Looking further, it does seem like the issue solved here is numpy 2 specific and other problems are tangential. So I'll resolve this to go in as-is

Reply:
thank you. though I'm not sure I understand.
The issue with the short matching of parentheses is not Numpy 2 specific. The modified test using tuple fails on older numpy versions as well. I can split the fix into a separate PR if that's preferable.

Was the suggestion to change the test to include tuple, but not include the fix and mark the test xfail instead, or leave the test as is and mark it xfail on numpy 2 only?

IIUC, there's a second issue with the test forwarding copies of globals() and locals() in `get_loaded_model_state`, as the model should not depend on what is loaded by the test. This makes the calls to np.float work even though they shouldn't. Is that a correct understanding? I can open a new issue for that.

Reply:
I think I was getting mixed up trying to replicate the tuple-mod failure on numpy<2 and couldn't. But I am able to now and see it's independent. I don't think it's necessary to split it because it's also required for numpy 2, but it can be. 

The original suggestion was to leave the test as-is and mark it as xfail because I didn't know how deep was the second issue, which you have the right understanding of. But I think it's pretty limited, and I have a small fix for it, so I think there's no need to xfail. I just hadn't figured a way to test the second issue currently, because I think it's peculiar to the awkward way `SampleIterator` gets saved into mdf as a string, which happens to include `"np."` in numpy 2.